kartothek/io/iter.py

```diff
@@ -68,33 +103,62 @@ def read_dataset_as_metapartitions__iterator(
         dispatch_by=dispatch_by,
     )
 
-    for mp in mps:
-        if concat_partitions_on_primary_index or dispatch_by:
-            mp = MetaPartition.concat_metapartitions(
-                [
-                    mp_inner.load_dataframes(
-                        store=store,
-                        tables=tables,
-                        columns=columns,
-                        categoricals=categoricals,
-                        predicate_pushdown_to_io=predicate_pushdown_to_io,
-                        predicates=predicates,
-                    )
-                    for mp_inner in mp
-                ]
-            )
-        else:
-            mp = cast(MetaPartition, mp)
-            mp = mp.load_dataframes(
-                store=store,
-                tables=tables,
-                columns=columns,
-                categoricals=categoricals,
-                predicate_pushdown_to_io=predicate_pushdown_to_io,
-                dates_as_object=dates_as_object,
-                predicates=predicates,
-            )
-        yield mp
+    if nthreads != 0:
+        mps = list(mps)
+        # this variable is currently not very usefull but one could introduce a Semaphore to controll the maximum amount of allowed partitions in memory. Probably not required, though.
+        if max_concurrent_results is None:
+            max_concurrent_results = 1
+        result_queue = Queue(max_concurrent_results)
+        store_pool = StorePool(ds_factory.store_factory)
+        fetch_partition = partial(
+            _process_job,
+            queue=result_queue,
+            method=MetaPartition.load_dataframes,
+            store_pool=store_pool,
+            tables=tables,
+            columns=columns,
+            categoricals=categoricals,
+            predicate_pushdown_to_io=predicate_pushdown_to_io,
+            dates_as_object=dates_as_object,
+            predicates=predicates,
+        )
+        with ThreadPoolExecutor(nthreads) as pool:
+            pool.map(fetch_partition, mps)
+            # It is very important that this loop is within the threadpool ctx mngr. Otherwise all computations are calculated before anything is yielded to be able to shut down the executor
+            print("enter loop", flush=True)
+            # must not iterate over the futures object since this may block
+            for _ in range(len(mps)):
+                yield result_queue.get()
+                print("yielded", flush=True)
+            print("done", flush=True)
+    else:
```

Comment:
What would be the overhead/dowside of using a single code path, i.e., spawning one thread in case no parallelism is wanted?

Reply:
this implementation implies that there are up to (1 + nthreads) partitions in memory. If the iter interface is used in situations where the memory footprint is supposed to be low this might be undesired.
I'll add this to the docs, of course